src/divergence_block_chain.cc

```diff
@@ -0,0 +1,177 @@
+#include "divergence_block_chain.h"
+
+namespace celerity::detail::divergence_checker_detail {
+
+void divergence_block_chain::add_new_hashes() {
+	std::lock_guard<std::mutex> lock(m_task_records_mutex);
+	for(size_t i = m_hashes_added; i < m_task_records.size(); ++i) {
+		std::size_t seed = m_local_hashes.empty() ? 0 : m_local_hashes.back();
+		celerity::detail::utils::hash_combine(seed, std::hash<task_record>{}(m_task_records[i]));
+		m_local_hashes.push_back(seed);
+	}
+	m_hashes_added = m_task_records.size();
+}
+
+void divergence_block_chain::clear(const int min_progress) {
+	m_local_hashes.erase(m_local_hashes.begin(), m_local_hashes.begin() + min_progress);
+	m_tasks_checked += min_progress;
+
+	m_last_cleared = std::chrono::steady_clock::now();
+}
+
+std::pair<int, int> divergence_block_chain::collect_hash_counts() {
+	m_per_node_hash_counts[m_local_nid] = static_cast<int>(m_local_hashes.size());
+
+	m_communicator->allgather_inplace(m_per_node_hash_counts.data(), 1);
+
+	const auto [min, max] = std::minmax_element(m_per_node_hash_counts.cbegin(), m_per_node_hash_counts.cend());
+
+	return {*min, *max};
+}
+
+per_node_task_hashes divergence_block_chain::collect_hashes(const int min_hash_count) const {
+	per_node_task_hashes data(min_hash_count, m_num_nodes);
+
+	m_communicator->allgather(m_local_hashes.data(), min_hash_count, data.data(), min_hash_count);
+
+	return data;
+}
+
+
+divergence_map divergence_block_chain::create_check_map(const per_node_task_hashes& task_hashes, const int task_num) const {
+	divergence_map check_map;
+	for(size_t i = 0; i < m_num_nodes; ++i) {
+		check_map[task_hashes(i, task_num)].push_back(i);
+	}
+	return check_map;
+}
+
+void divergence_block_chain::check_for_deadlock() const {
+	auto diff = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::steady_clock::now() - m_last_cleared);
+	static auto last = std::chrono::seconds(0);
+
+	if(diff >= std::chrono::seconds(10) && diff - last >= std::chrono::seconds(5)) {
+		std::string warning = fmt::format("After {} seconds of waiting nodes", diff.count());
+
+		for(size_t i = 0; i < m_num_nodes; ++i) {
+			if(m_per_node_hash_counts[i] == 0) { warning += fmt::format(" {},", i); }
+		}
+
+		warning += " did not move to the next task. The runtime might be stuck.";
+
+		CELERITY_WARN("{}", warning);
+		last = diff;
+	}
+}
+
+void divergence_block_chain::log_node_divergences(const divergence_map& check_map, const int task_num) {
+	std::string error = fmt::format("Divergence detected in task graph at index {}:\n\n", task_num);
```

Comment:
As also discussed in-person I reformulated the whole error message. Now we get an error message like this one:

```
[2023-12-18 11:50:13.109] [0] [error] Divergence detected. Task Nr 4 diverges on nodes:

Following task-hash 0xaa7e7279ec138abe resulted on node  2 
Following task-hash 0x816fa348a1f52942 resulted on node  1 
Following task-hash 0x1fecf12dff1bff6c resulted on nodes 0 3 

[2023-12-18 11:50:13.109] [0] [error] Task record for hash 0x1fecf12dff1bff6c:

id: 4, debug_name: one_third_kernel, type: device-compute, cgid: 0
geometry:
         dimensions: 1, global_size: [10000,1,1], global_offset: [0,0,0], granularity: [1,1,1]
accesses: 
         bid: 0, buffer_name: B0 "mat_a", mode: 3, req: {[0,0,0] - [10000,1,1]}
dependencies: 
         node: 3, kind: anti-dep, origin: dataflow
         node: 0, kind: true-dep, origin: last-epoch
```

This includes now the `task_id` instead of `task_num`, which is much more informative.